File: src/os.js
import {
    keyTypes,
    normalizeKeyValue
} from "./keys.js";

const gesture = Object.seal({
    type: "",
    text: "",
    command: ""
});

// Translates operating system-specific Browser KeyboardEvents into a
// cross-platform Gesture that can then be dispatched to a CommandPack
// for translation to an EditorCommand.
class OperatingSystem {
    constructor(osName, pre1, pre2, redo, pre3, home, end, pre5, fullHome, fullEnd) {
        this.name = osName;

        const pre4 = pre3;
        pre3 = pre3.length > 0 ? pre3 : "Normal";

        const definitions = Object.freeze(new Map([
            ["CursorDown", "Normal_ArrowDown"],
            ["CursorLeft", "Normal_ArrowLeft"],
            ["CursorRight", "Normal_ArrowRight"],
            ["CursorUp", "Normal_ArrowUp"],
            ["CursorPageDown", "Normal_PageDown"],
            ["CursorPageUp", "Normal_PageUp"],
            ["CursorSkipLeft", `${pre2}_ArrowLeft`],
            ["CursorSkipRight", `${pre2}_ArrowRight`],
            ["CursorHome", `${pre3}_${home}`],
            ["CursorEnd", `${pre3}_${end}`],
            ["CursorFullHome", `${pre5}_${fullHome}`],
            ["CursorFullEnd", `${pre5}_${fullEnd}`],


            ["SelectDown", "Shift_ArrowDown"],
            ["SelectLeft", "Shift_ArrowLeft"],
            ["SelectRight", "Shift_ArrowRight"],
            ["SelectUp", "Shift_ArrowUp"],
            ["SelectPageDown", "Shift_PageDown"],
            ["SelectPageUp", "Shift_PageUp"],
            ["SelectSkipLeft", `${pre2}Shift_ArrowLeft`],
            ["SelectSkipRight", `${pre2}Shift_ArrowRight`],
            ["SelectHome", `${pre4}Shift_${home}`],
            ["SelectEnd", `${pre4}Shift_${end}`],
            ["SelectFullHome", `${pre5}Shift_${fullHome}`],
            ["SelectFullEnd", `${pre5}Shift_${fullEnd}`],

            ["SelectAll", `${pre1}_a`],

            ["ScrollDown", `${pre1}_ArrowDown`],
            ["ScrollUp", `${pre1}_ArrowUp`],

            ["DeleteLetterLeft", "Normal_Backspace"],
            ["DeleteLetterRight", "Normal_Delete"],
            ["DeleteWordLeft", `${pre2}_Backspace`],
            ["DeleteWordRight", `${pre2}_Delete`],
            ["DeleteLine", `Shift_Delete`],

            ["AppendNewline", "Normal_Enter"],
            ["PrependNewline", `${pre2}_Enter`],

            ["InsertTab", "Normal_Tab"],
            ["RemoveTab", "Shift_Tab"],

            ["Undo", `${pre1}_z`],
            ["Redo", redo]
        ]));

        const substitutions = new Map();
        for (let pair of definitions) {
            substitutions.set(pair[1], pair[0]);
        }

        this.makeCommand = (evt) => {
            gesture.text = normalizeKeyValue(evt);

            gesture.type = keyTypes.has(gesture.text)
                ? keyTypes.get(gesture.text)
                : "printable";

            gesture.command = "";

            if (evt.ctrlKey
                || evt.altKey
                || evt.metaKey) {
                if (gesture.type === "printable"
                    || gesture.type === "whitespace") {
                    gesture.type = "special";
                }

                if (evt.ctrlKey) {
                    gesture.command += "Control";
                }

                if (evt.altKey) {
                    gesture.command += "Alt";
                }

                if (evt.metaKey) {
                    gesture.command += "Meta";
                }
            }

            if (evt.shiftKey) {
                gesture.command += "Shift";
            }

            if (gesture.command === "") {
                gesture.command += "Normal";
            }

            gesture.command += "_" + gesture.text;

            if (substitutions.has(gesture.command)) {
                gesture.command = substitutions.get(gesture.command);
            }

            return gesture;
        }

        Object.freeze(this);
    }
};

export const Windows = new OperatingSystem(
    "Windows",
    "Control", "Control",
    "Control_y",
    "", "Home", "End",
    "Control", "Home", "End");

export const MacOS = new OperatingSystem(
    "macOS",
    "Meta", "Alt",
    "MetaShift_z",
    "Meta", "ArrowLeft", "ArrowRight",
    "Meta", "ArrowUp", "ArrowDown");

File: csharp/Primrose.NETStandard2/OS.cs
using System.Collections.Generic;

namespace Primrose
{
    /// <summary>
    /// Translates operating system-specific Browser KeyboardEvents into a
    /// cross-platform Gesture that can then be dispatched to a CommandPack
    /// for translation to an EditorCommand.
    /// </summary>
    public class OperatingSystem
    {
        public static readonly OperatingSystem Windows = new OperatingSystem(
            "Windows",
            "Control", "Control",
            "Control_y",
            "", "Home", "End",
            "Control", "Home", "End");

        public static readonly OperatingSystem MacOS = new OperatingSystem(
            "macOS",
            "Meta", "Alt",
            "MetaShift_z",
            "Meta", "ArrowLeft", "ArrowRight",
            "Meta", "ArrowUp", "ArrowDown");

        public string name { get; }
        private readonly Dictionary<string, string> substitutions;

        public OperatingSystem(string osName, string pre1, string pre2, string redo, string pre3, string home, string end, string pre5, string fullHome, string fullEnd)
        {
            name = osName;

            var pre4 = pre3;
            if (pre3.Length == 0)
            {
                pre3 = "Normal";
            }

            substitutions = new Dictionary<string, string>() {
                { "Normal_ArrowDown", "CursorDown" },
                { "Normal_ArrowLeft", "CursorLeft" },
                { "Normal_ArrowRight", "CursorRight" },
                { "Normal_ArrowUp", "CursorUp" },
                { "Normal_PageDown", "CursorPageDown" },
                { "Normal_PageUp", "CursorPageUp" },
                { $"{pre2}_ArrowLeft", "CursorSkipLeft" },
                { $"{pre2}_ArrowRight", "CursorSkipRight" },
                { $"{pre3}_{home}", "CursorHome" },
                { $"{pre3}_{end}", "CursorEnd" },
                { $"{pre5}_{fullHome}", "CursorFullHome" },
                { $"{pre5}_{fullEnd}", "CursorFullEnd" },

                { "Shift_ArrowDown", "SelectDown" },
                { "Shift_ArrowLeft", "SelectLeft" },
                { "Shift_ArrowRight", "SelectRight" },
                { "Shift_ArrowUp", "SelectUp" },
                { "Shift_PageDown", "SelectPageDown" },
                { "Shift_PageUp", "SelectPageUp" },
                { $"{pre2}Shift_ArrowLeft", "SelectSkipLeft" },
                { $"{pre2}Shift_ArrowRight", "SelectSkipRight" },
                { $"{pre4}Shift_{home}", "SelectHome" },
                { $"{pre4}Shift_{end}", "SelectEnd" },
                { $"{pre5}Shift_{fullHome}", "SelectFullHome" },
                { $"{pre5}Shift_{fullEnd}", "SelectFullEnd" },

                { $"{pre1}_a", "SelectAll" },

                { $"{pre1}_ArrowDown", "ScrollDown" },
                { $"{pre1}_ArrowUp", "ScrollUp" },

                { "Normal_Backspace", "DeleteLetterLeft" },
                { "Normal_Delete", "DeleteLetterRight" },
                { $"{pre2}_Backspace", "DeleteWordLeft" },
                { $"{pre2}_Delete", "DeleteWordRight" },
                { "Shift_Delete", "DeleteLine" },

                { "Normal_Enter", "AppendNewline" },
                { $"{pre2}_Enter", "PrependNewline" },

                { "Normal_Tab", "InsertTab" },
                { "Shift_Tab", "RemoveTab" },

                { $"{pre1}_z", "Undo" },
                { redo, "Redo" }
            };
        }

        public Gesture makeCommand(KeyEvent evt)
        {
            var gesture = new Gesture();
            gesture.text = Keys.normalizeKeyValue(evt);
            gesture.type = Keys.keyTypes.ContainsKey(gesture.text)
                ? Keys.keyTypes[gesture.text]
                : "printable";

            if (evt.ctrlKey
                || evt.altKey
                || evt.metaKey)
            {
                if (gesture.type == "printable"
                    || gesture.type == "whitespace")
                {
                    gesture.type = "special";
                }

                if (evt.ctrlKey)
                {
                    gesture.command += "Control";
                }

                if (evt.altKey)
                {
                    gesture.command += "Alt";
                }

                if (evt.metaKey)
                {
                    gesture.command += "Meta";
                }
            }

            if (evt.shiftKey)
            {
                gesture.command += "Shift";
            }

            if (gesture.command == "")
            {
                gesture.command += "Normal";
            }

            gesture.command += "_" + gesture.text;

            if (substitutions.ContainsKey(gesture.command))
            {
                gesture.command = substitutions[gesture.command];
            }

            if (gesture.command == "PrependNewline")
            {
                gesture.type = "whitespace";
            }

            return gesture;
        }
    }
}